sum method for atomgroups

PicoCentauri opened the pull request:
Changes made in this Pull Request:
 -  Added `*Group.sum()` method
 - Added compound parameter to `total_mass` and `total_charge`

Following the discussion of PR #2118 this PR adds a sum method to all groups. Furthermore a compound parameter to `*Group.total_charge()` and `*Group.total_mass()` was added similar to the `center_of_mass()`  method. 

I'm not sure if the name `sum`  is confusing but this was the best thing, which came to my mind.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

richardjgowers commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,115 @@ def center_of_geometry(self, pbc=None, compound='group'):
+            >>> sel = u.select_atoms('prop mass > 4.0')
+            >>> sel.sum(sel.charges)
+
+        To find the total mass per residue of all CA :class:`Atoms<Atom>`::
```
I'm confused what this will return.  Is it the mass of CA atoms aggregated per residue?  So if each residue only had one CA it would be identical across all?

PicoCentauri replied:
> Is it the mass of CA atoms aggregated per residue?

Yes, accectly.

> So if each residue only had one CA it would be identical across all?

Yes. In the case for only one residue this is a bit senseless. But could be useful if you have for example a geometric selection...

richardjgowers commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,115 @@ def center_of_geometry(self, pbc=None, compound='group'):
+        To find the total charge of a given :class:`AtomGroup`::
+
+            >>> sel = u.select_atoms('prop mass > 4.0')
+            >>> sel.sum(sel.charges)
```
This seems clunky.  How tricky would it be making `sel.sum('charges')` work?  But really `sel.charges.sum()` seems most preferable

PicoCentauri replied:
`sel.charges.sum()` is nice, but could this brake some scripts where people use the numpy attribites of the `sum` method?

richardjgowers commented:
I think I get what this is doing, but maybe it would be better to achieve it using some sort of `groupby` feature instead.

So `sel.groupby('residues').charges`, `sel.groupby('fragment').total_mass()`

So `sel.groupby('residues')` goes to a ResidueGroup `sel.groupby('fragment')` goes to an iterable of AtomGroups

@MDAnalysis/coredevs thoughts?

zemanj commented:
I generally like the idea of having the opportunity of summing up group properties per compound.
However, I'd prefer it if the `*Group.sum()` method accepted the name of a property (as a string) instead of an array containing the values.
I don't know if one really needs the flexibility of passing arbitrary values. You could have both if you check for the type of the `elements` argument (which I would rename to `property`) and use `getattr()` in case of a string and otherwise, try to use the values as is.

What you're missing right now is a check for the length of the elements array, which has to  match the number of atoms.

And there's a typo in the return statements of `total_mass()` and `total_charge()` (it must read `elements`, not `weights`).

zemanj commented:
@richardjgowers I'm afraid `*Group.groupby()` returns a dict containing groups as values, which is dead slow because of all the object instantiations.

jbarnoud commented:
I do not like the name. I do not like that `sum(group)` does not do the same thing as `group.sum()`. I would not mind a `*Group.sum_per_compound` method, though (or any similar name).

For now, I solve this kind of problems like:

```python
total_charges = np.array([compound.sum() for compound in sel.split('molecules')])
```

Though, what I really need most of the time is:

```python
all_centers = np.array([compound.center_of_mass() for compound in sel.split('molecules')])
```

PicoCentauri commented:
@zemanj 
> I'd prefer it if the *Group.sum() method accepted the name of a property (as a string) instead of an array containing the values.

This should be no problem to change.

@jbarnoud 
Both are nice pieces of code, and do exactly the same as `toal_charge('molecules')`  and `center_of_mass('molecules')` but probably not all users are aware of this. I think it would be nice to have a similar basis for calculating grouped centers (`center_of_mass` etc) and grouped sums of properties (`total_mass` etc). One could kick out the sum method in every group and just add some logic to the` total_charge` and `total_mass` methods?

zemanj commented:
@jbarnoud I just checked the runtimes of your approach against `center_of_mass(compound='molecules')`:
```Python
In [1]: import MDAnalysis as mda
In [2]: import numpy as np
In [3]: u = mda.Universe("topol.tpr", "traj_comp.xtc");
In [4]: ag = u.select_atoms("moltype BMIM")
In [5]: ag
Out[5]: <AtomGroup with 12500 atoms>
In [6]: len(np.unique(ag.molnums))
Out[6]: 500
In [7]: %timeit np.array([compound.center_of_mass(pbc=True) for compound in ag.split('molecule')])
10 loops, best of 3: 33.1 ms per loop
In [8]: %timeit ag.center_of_mass(pbc=True, compound='molecules')
1000 loops, best of 3: 1.23 ms per loop
```
Your approach takes about **27** times longer than `ag.center_of_mass(compound='molecules')`.
The difference in runtime is 31.87 ms, which amounts to **53 min** when analyzing 100,000 frames.
That certainly justifies a faster implementation!

zemanj commented:
@jbarnoud
> I do not like the name. I do not like that `sum(group)` does not do the same thing as `group.sum()`. I would not mind a `*Group.sum_per_compound` method, though (or any similar name).

I agree that the name `sum()` is sub-optimal. But why not follow a more flexible approach and do something like `*Group.accumulate_properties(property, compound, accumulator=np.sum)`?
That way, one could pass arbitrary functions as accumulators.

orbeckst commented:
Would this be a `Group.reduce(..., func=np.sum)`? Using `min`, `max`, `median`, `std`, `avg` will make it easy to do all kinds of interesting things.

For the most common ones I would still be tempted to create shorthands `Group.sum()`, `Group.avg()` etc – makes TAB completion and interactive exploration much more intuitive.

orbeckst commented:
(Sorry, I ignored the discussion regarding the appropriateness of the name `sum` – I agree with https://github.com/MDAnalysis/mdanalysis/pull/2192#issuecomment-459407520 that `sum(group)` should do the same as `group.sum()`.)

jbarnoud commented:
@zemanj Thank you for the benckmark. I did not notice the compound argument in `center_of_mass`. Since the argument already exists, it makes perfect sense to have it everywhere it is relevant; `total_mass` and `total_charges` are perfect candidates here.

I like the accumulator/reduce idea. Because the proposed function can already work with any array, it opens very interesting possibilities. I believe the function would work with any array assuming that the first dimension matches the number of atoms; am I right here?

PicoCentauri commented:
To sum up the discussion:
The idea is to have a function like `*group.accumulate(attribute, op=np.sum, compound='group')`, that accumulates the given attribute (mass, charge) of the group according to the given operation?

PicoCentauri commented:
I just adjusted the method and also did a first performance test comparing it to list comprehension.

```python
In [9]: import MDAnalysis as mda 
   ...: from MDAnalysisTests.datafiles import PSF, DCD 
   ...: u = mda.Universe(PSF,DCD) 
   ...: %timeit u.atoms.accumulate("masses",compound="residues", op=np.sum) 
   ...: %timeit np.array([np.sum(compound.masses) for compound in u.atoms.split('residue')])                                                                                                             

1.41 ms ± 33.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
4.04 ms ± 30.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

jbarnoud commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,128 @@ def center_of_geometry(self, pbc=None, compound='group'):
+
+        Parameters
+        ----------
+        attribute : str
```
I liked that you could give an arbitrary array (assuming it has the right size). In the ideal case, I would like to be able to provide either the name, or an arbitrary array.

PicoCentauri replied:
Yes. This should be feasible.

jbarnoud commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,128 @@ def center_of_geometry(self, pbc=None, compound='group'):
+
+        Returns
+        -------
+        sum : numpy.ndarray
```
You do not have to provide a name for the return value. Here I would avoid doing it as the name is actually wrong:

```
Returns
-------
numpy.ndarray
    The description
```

jbarnoud commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,128 @@ def center_of_geometry(self, pbc=None, compound='group'):
+            compound_mask = compound_sizes == compound_size
+            _compound_indices = unique_compound_indices[compound_mask]
+            atoms_mask = np.in1d(compound_indices, _compound_indices)
+            _summands = attribute_values[atoms_mask].reshape((-1, compound_size))
```
What does "summands" mean as a variable name?

PicoCentauri replied:
These are the elements to which the accumulation is applied. I will rename them to elements if this is more clear.

PicoCentauri commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,128 @@ def center_of_geometry(self, pbc=None, compound='group'):
+        ----------
+        attribute : str
+            Attribute to accumulate.
+        op : operation, optional
```
Okay I will improve the docs. Another constraint to the function currently is that it also should be fine with the `axis=1` option.

PicoCentauri commented:
I added some tests and also the possibility to pass multi-dimensional arrays.

zemanj commented:
@PicoCentauri According to CodeCov, additional tests for `compound='molecules'` and for universes without a) `molnums` and b) `bonds` are needed. Apart from that, this looks really good :+1:

PicoCentauri commented:
@zemanj Is there anything open here? Otherwise would be nice if we could merge this so that I can continue the work on PR #2118.

zemanj commented:
@PicoCentauri Thanks for the reminder, I'll check it out later today.

zemanj commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,150 @@ def center_of_geometry(self, pbc=None, compound='group'):
+        compound : {'group', 'segments', 'residues', 'molecules', 'fragments'}, optional
+            If ``'group'``, the sum over all elements associated
+            with atoms in the group will
+            be returned as a single value. Else, the sum for all elements
```
```suggestion
            be returned as a single value. Otherwise, the accumulation of the attributes
```

zemanj commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,150 @@ def center_of_geometry(self, pbc=None, compound='group'):
+            If ``'group'``, the sum over all elements associated
+            with atoms in the group will
+            be returned as a single value. Else, the sum for all elements
+            of each :class:`Segment`, :class:`Residue`, molecule, or fragment
```
```suggestion
            per :class:`Segment`, :class:`Residue`, molecule, or fragment
```

zemanj commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,150 @@ def center_of_geometry(self, pbc=None, compound='group'):
+            If a :class:`numpy.ndarray` is provided which first dimensions 
+            must have the same length as the total number of atoms in 
+            the :class:`AtomGroup`.
+        function : function, optional
```
```suggestion
        function : callable, optional
```

PicoCentauri commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,150 @@ def center_of_geometry(self, pbc=None, compound='group'):
+            the :class:`AtomGroup`.
+        function : function, optional
+            Numpy function to perform. It must take an array of the same dtype 
+            and return a scalar.
```
Sorry for the bad docu here. You are right. Its not restricted to numpy functions and the dtype is also not important. The only thing it necessary is an axis argument as in the numpy functions. Currently I see no other smart way to implement this...

zemanj replied:
I don't see any better way either, so I think fixing the docs is sufficient. Something along these lines should do the trick:
```Python
"""
...
function : callable, optional
    The function performing the accumulation. It must take the array of
    attribute values to accumulate as its only positional argument and
    accept an (optional) keyword argument ``axis`` allowing to specify
    the axis along which the accumulation is performed.
...
"""
```
But maybe you have a more concise phrasing in mind.

PicoCentauri replied:
Sounds nice!

PicoCentauri commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,150 @@ def center_of_geometry(self, pbc=None, compound='group'):
+                             " one of 'group', 'residues', 'segments', "
+                             "'molecules', or 'fragments'.".format(compound))
+
+        higher_dimemsions = list(attribute_values.shape[1:])
```
Cheers :blush:

PicoCentauri commented on package/MDAnalysis/core/groups.py:
```diff
@@ -850,6 +850,150 @@ def center_of_geometry(self, pbc=None, compound='group'):
+                attribute_values = getattr(atoms, attribute)
+            except AttributeError:
+                raise NoDataError("AtomGroup does not have '{}' "
+                                  "attribute.".format(attribute))
```
Yes why not. Following this I can also remove this from the `Raises` section of the docstring and the specific test, right?

PicoCentauri commented on testsuite/MDAnalysisTests/core/test_atomgroup.py:
```diff
@@ -488,6 +488,64 @@ def test_center_wrong_shape(self, ag):
         with pytest.raises(ValueError):
             ag.center(weights)
 
+class TestAccumulate(object):
```
Thanks.

You are right. I will ad them. Should be doable using the awesome multiple fixtures of pytest. I would rather move all the tests to `core/test_groups.py`. In my opinion this is more consistent. Why should we create a new test file only for accumulation and not for the other group attributes?

The `shape` is implicitly tested in the `test_accumulate_array_attribute_compounds` function. This checks the full array and would raise a shapes missmatch. But if you want I can also add a specific shape test?

zemanj commented:
I'd suggest to remove it from the docstring but change the tests to check for an AttributeError. Technically, the test isn't required, but it might help us to keep track of this behavior if we change something in AtomGroup.__getattr__() or related.

On February 18, 2019 9:30:32 PM GMT+01:00, Philip Loche <notifications@github.com> wrote:
>PicoCentauri commented on this pull request.
>
>
>
>> +
>+        atoms = self.atoms
>+
>+        if type(attribute) in (np.ndarray, tuple, list):
>+            attribute_values = np.asarray(attribute)
>+            if len(attribute_values) != len(atoms):
>+                raise ValueError("The input array length {} does not
>match "
>+                                 "the required length {} based on "
>+                                 "the group
>size.".format(len(attribute_values),
>+                                                          len(atoms)))
>+        else:
>+            try:
>+                attribute_values = getattr(atoms, attribute)
>+            except AttributeError:
>+                raise NoDataError("AtomGroup does not have '{}' "
>+                                  "attribute.".format(attribute))
>
>Yes why not. Following this I can also remove this from the `Raises`
>section of the docstring and the specific test, right?  
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/MDAnalysis/mdanalysis/pull/2192#discussion_r257819907

zemanj commented:
@PicoCentauri
> The `shape` is implicitly tested in the `test_accumulate_array_attribute_compounds` function. This checks the full array and would raise a shapes missmatch. But if you want I can also add a specific shape test?

Looks like I missed that one. Would you mind changing the tested array's shape to `(len(ag), 2, 5)`? That will guarantee that the "remaining" shape won't simply be flattened.

PicoCentauri commented on testsuite/MDAnalysisTests/core/test_accumulate.py:
```diff
@@ -0,0 +1,165 @@
+
+universe = mda.Universe(PSF, DCD)
+universe_molfrag = mda.Universe(TPR_xvf, TRR_xvf)
+levellist = ('atoms', 'residues', 'segments')
```
I added the tests for segments; However, I am not happy with putting the decorator `@pytest.mark.parametrize('level', levellist)` in front of **every** function in the class but unfortunately I haven't found a better solution. Maybe somebody who knows pytest better finds a smarter way...

richardjgowers replied:
You can place the decorator at the class level, but then **every** test in the class needs to use it

```python
@pytest.mark.parametrize('this', [1, 2])
class TestThing(object):
    def test_add(self, this):
        something
```

Or you can make a fixture have a loop in it using params

```python
@pytest.fixture(params=[1, 2])
def thing(request):
    return request.param[0]

def test_this(thing):
    assert thing == 1 or thing == 2
```

PicoCentauri replied:
Nice. Thanks your first suggestion will do the job here.

richardjgowers commented on testsuite/MDAnalysisTests/core/test_accumulate.py:
```diff
@@ -0,0 +1,165 @@
+from MDAnalysisTests.datafiles import PSF, DCD, TPR_xvf, TRR_xvf, GRO
+import pytest
+
+universe = mda.Universe(PSF, DCD)
```
you can't have these objects live here.  This might mean you have to expand a few test functions below too unfortunately

PicoCentauri replied:
Okay... :worried:

richardjgowers commented on testsuite/MDAnalysisTests/core/test_accumulate.py:
```diff
@@ -0,0 +1,165 @@
+    """Tests the functionality of *Group.accumulate()."""
+
+    @pytest.mark.parametrize('level', levellist)
+    def test_accumulate_str_attribute(self, level):
```
so here you could do something like:

```python
@pytest.fixture(params['atoms', 'residues' 'segments'])
def group(request):
    u = mda.Universe(..., ...)
    return getattr(u, request.params[0])

def test_accumulate_str_attribute(group):
    assert_almost_equal(group.accumulate(), np.sum())
```
and the fixture params make the test expand to cover all 3 types of Group

richardjgowers commented on testsuite/MDAnalysisTests/core/test_accumulate.py:
```diff
@@ -0,0 +1,144 @@
+            np.prod(group.atoms.masses))
+
+    @pytest.mark.parametrize('universe, name, compound',
+                             ((mda.Universe(PSF, DCD), 'resids', 'residues'),
```
this has the same problem, the scope of the Universe objects here will be the entire test run.  The advantage of fixtures is the object only lives as long as the test function

zemanj replied:
I just replaced the DCD / TPR universes with the infamous `UnWrapUniverse` for the compound tests. Creating `UnWrapUniverse`s is so fast that fixtures won't improve anything here.
Hope you're ok with this.

zemanj commented:
@PicoCentauri I found a couple of typos in the docstrings which I missed in my review. I ddn't want to bug you with this again, so I fixed it myself. I also changed the `attribute` instance check to test for string instances first, which should be the normal use case I think.
Furthermore, I used the `UnWrapUniverse` from `MDAnalysisTests.core.util` for the compound tests. That saves a couple of seconds and doesn't require fixtures.

PicoCentauri commented:
@zemanj Thanks for fixing the typos. I'm happy with the tests and the method.

richardjgowers commented:
Awesome, more commits on the board for @PicoCentauri !

PicoCentauri commented:
Awesome. Thanks for the help and for merging!
